### PR TITLE
Add handler for serving language redirections based on preferred locale

### PR DESCRIPTION
**Pull request by devnook**
Add a handler that redirects the request based on requested locale, according to priority:
 1. Locale directly expressed in the url, e.g. /en/some-post
 2. Locale in accept-language HTTP header.
 3. Locale in "preferred_locale" cookie.
 If requested locale's content does not exist, falls back to default ('en').

Note for testing:
This PR makes more sense when combined with the PR #2123 as you need some "localised" content, e.g. /pl/ directory to test the handler.

### CODE REVIEW

File: locale-handler.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+const DEFAULT_LOCALE = "en";
+
+const isSupported = (locale) => SUPPORTED_LOCALES.indexOf(locale) > -1;
+
+/**
+ * A handler that redirects the request based on requested locale,
+ * according to priority:
+ * 1. Locale directly expressed in the url, e.g. /en/some-post
+ * 2. Locale in accept-language HTTP header.
+ * 3. Locale in "preferrec_locale" cookie.
```
Comment (samthor) [inline locale-handler.js]: This should be "preferred_lang"

File: locale-handler.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * If requested locale's content does not exist, falls back to default.
+ * @param {Object} req Request object.
+ * @param {Object} res Response object.
+ * @param {Function} next Middleware next handler.
+ * @return {!Function}
+ */
+module.exports = (req, res, next) => {
+  // Check if url is navigational.
+  if (req.url.endsWith("/")) {
```
Comment (samthor) [inline locale-handler.js]: style nit: this block always returns, and then we have a tiny alternative case below (`return next();`). Can we flip this around?

```js
  if (!req.url.endsWith("/")) {
    return next();
  }
  // less indented code
```

File: locale-handler.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    if (lang === DEFAULT_LOCALE) {
+      // If this is alread default language, continue.
+      return next();
+    }
+
+    const localizedFilePath = path.join(
+      __dirname,
+      "dist", // Must serve from dist directory even in dev mode.
+      lang,
```
Comment (samthor) [inline locale-handler.js]: `lang` is untrusted. I could set my `preferred_lang` cookie to "../../escape/our/root/" and Node will join it.

Comment (robdodson) [reply locale-handler.js]: are you thinking it should be sanitized with a `isSupported()` check?

File: locale-handler.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  if (req.url.endsWith("/")) {
+    // Check if language is specified in the url.
+    const pathSegment = req.path.split("/")[1];
+    const isLangInPath = isSupported(pathSegment);
+    let lang;
+    let filePath;
+
+    if (isLangInPath) {
+      lang = pathSegment;
+      const pathParts = req.path.split("/");
```
[the remark this answers is not in the document]
Comment (devnook) [reply locale-handler.js]: done

File: locale-handler.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const fs = require("fs");
+const localeCode = require("iso-639-1");
+const path = require("path");
+
+const isProd = Boolean(process.env.GAE_APPLICATION);
+const contentDir = isProd ? "dist" : "src/site/content";
```
[the remark this answers is not in the document]
Comment (devnook) [reply locale-handler.js]: At this point of build process /dist might not be populated yet.

File: locale-handler.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    const localizedFilePath = path.join(
+      __dirname,
+      "dist", // Must serve from dist directory even in dev mode.
+      lang,
+      filePath,
+      "index.html",
+    );
+
+    if (fs.existsSync(localizedFilePath)) {
+      return isLangInPath ? next() : res.redirect(`/${lang}${filePath}`);
```
[the remark this answers is not in the document]
Comment (devnook) [reply locale-handler.js]: done

File: locale-handler.js (hunk 2 above)
Comment (devnook) [reply locale-handler.js]: done

File: locale-handler.js (hunk 1 above)
Comment (devnook) [reply locale-handler.js]: done

File: locale-handler.js (hunk 3 above)
Comment (devnook) [reply locale-handler.js]: I whitelist allowed values in isSupported. Do you think there should be some additional check here?

File: locale-handler.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (robdodson) [reply locale-handler.js]: I'm not sure I understand why? I don't think we pull from `src/` for any other part of the server do we? I thought we usually ran `build` before starting the server.

File: locale-handler.js (hunk 3 above)
Comment (robdodson) [reply locale-handler.js]: I see. It looks like line 57 sanitizes this value I think.

File: locale-handler.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (devnook) [reply locale-handler.js]: I based this assumption on this comment in server.js: https://github.com/GoogleChrome/web.dev/blob/master/server.js#L24  I think there is no guarantee that build runs before the server starts (e.g. if you run npm run start in stead of npm run dev)

**Pull request merged**